File: 1-two-sum/two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        
        HashMap<Integer, Integer> twoSum = new HashMap<>();
        // declare a HashMap with key and value as Integer type

        for(int i=0; i<nums.length; i++){
            int diff = target - nums[i];

            if(twoSum.containsKey(diff) && twoSum.get(diff) != i){
            // twoSum.containsKey(diff) checks if the key (number) is present in the map
            // twoSum.get(diff) checks if the diff key index and the current index are different
            // diff and i cannot have the same index value

                return new int[]{twoSum.get(diff), i};
                // returns an array containing the indexes of the keys that add up to target
            }
            twoSum.put(nums[i], i);
            // stores key-value pairs in the map
            // here the keys are the numbers and the values are their indices
        }

        return new int[]{};
    }
}

File: 1-two-sum/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        
        twoSum = {}
        # empty dictionary to store key->value pairs
        # the keys are the numbers and the values are their indices

        for i,n in enumerate(nums):
            twoSum[n] = i
            # i  = index
            # n = value at that index
            # will return a dictionary with key->value pairs
            # for example: [1,2,3] will return {1:0, 2:1, 3:2}

        for i,n in enumerate(nums):
            diff = target - n
            if diff in twoSum and twoSum[diff]!=i:
                return [twoSum[diff], i]